File: Leetcode - Python/1120_tree_maximumAverageSubtree.py
# https://leetcode.com/problems/maximum-average-subtree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maximumAverageSubtree(self, root: TreeNode) -> float:
        self.res = 0
        def calcAve(root):
            if not root:
                return [0, 0.0]
            n1, s1 = calcAve(root.left)
            n2, s2 = calcAve(root.right)
            n = n1 + n2 + 1
            s = s1 + s2 + root.val
            self.res = max(self.res, s/n)
            return [n,s]
        calcAve(root)
        return self.res

File: Leetcode - Java/1120_tree_maximumAverageSubtree.java
// https://leetcode.com/problems/maximum-average-subtree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private double res = 0;
    
    public double maximumAverageSubtree(TreeNode root) {
        calcAve(root);
        return res;
    }
    
    private double[] calcAve(TreeNode root) {
        if (root == null) {
            return new double[]{0, 0.0};
        }
        double[] l = calcAve(root.left), r = calcAve(root.right);
        double sum = l[0] + r[0] + root.val;
        double n = l[1] + r[1] + 1;
        
        res = Math.max(res, sum/n);
        return new double[]{sum,n};
    }
}